### PR TITLE
[SPARK-23312][SQL][followup] add a config to turn off vectorized cache reader

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

https://github.com/apache/spark/pull/20483 tried to provide a way to turn off the new columnar cache reader, to restore the behavior in 2.2. However even we turn off that config, the behavior is still different than 2.2.

If the output data are rows, we still enable whole stage codegen for the scan node, which is different with 2.2, we should also fix it.

## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
@sameeragarwal @kiszk  @viirya @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 1]
```diff
@@ -61,6 +61,9 @@ case class InMemoryTableScanExec(
     }) && !WholeStageCodegenExec.isTooManyFields(conf, relation.schema)
   }
 
+  // TODO: revisit this. Shall we always turn off whole stage codegen if the output data are rows?
+  override def supportCodegen: Boolean = supportsBatch
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: In 2.4 we should look into this. My gut feeling is we don't need to enable whole stage codegen for scan nodes that output data as rows.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: Yeah, we can do more perf measurement after 2.3 release

**Comment by gatorsmile** [comment]
LGTM pending Jenkins.

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: I think this is safe to keep the same behavior of 2.2.

I'm not sure if enabling whole stage codegen can hurt performance for scan nodes, btw. We can revisit this, of course.

**Comment by viirya** [comment]
retest this please.

**Review by viirya — APPROVED** [review]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala (hunk 1 above)
Comment (kiszk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: We confirmed that we got performance improvements in several cases.
I will revisit this with more cases after 2.3 release.

**Comment by kiszk** [comment]
LGTM

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
```
java.lang.RuntimeException: [unresolved dependency: com.sun.jersey#jersey-json;1.14: configuration not found in com.sun.jersey#jersey-json;1.14: 'master(compile)'. Missing configuration: 'compile'. It was required from org.apache.hadoop#hadoop-yarn-common;2.6.5 compile]
```

Seems like some environmental problems, let's try again.

**Comment by cloud-fan** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master/2.3

**Pull request closed without merging**